lib/Importer/ProtobufLoader.cpp

```diff
@@ -73,12 +73,50 @@ ProtobufLoader::getStaticPlaceholderByNameOrNull(llvm::StringRef name) const {
   return (res && res->isStatic()) ? res : nullptr;
 }
 
+bool ProtobufLoader::isOpRequestedInFP16Precision(llvm::StringRef name) const {
+  auto opPrecisionUpdateItr =
+      std::find(operatorPrecisionUpdateInputOutputNames_.begin(),
+                operatorPrecisionUpdateInputOutputNames_.end(), name);
+  return opPrecisionUpdateItr != operatorPrecisionUpdateInputOutputNames_.end();
+}
+
+void ProtobufLoader::updateNodeOutputPrecision(llvm::StringRef outputName,
+                                               NodeValue NV) {
+  // If output name is specified to run in FP16 precision and output type
+  // is fp16 add convert to fp32 node and update nodeValueByName_ map,
+  // so that next operator in the model receives appropriate nodeValue.
+  if (isOpRequestedInFP16Precision(outputName) &&
+      NV.getElementType() == ElemKind::Float16Ty) {
+    std::string opName = outputName.str() + "_convToFP32";
+    auto *CT = G_->createConvertTo(opName, NV, ElemKind::FloatTy);
+    nodeValueByName_[outputName] = CT->getResult();
+  } else {
+    nodeValueByName_[outputName] = NV;
+  }
+}
+
 Constant *ProtobufLoader::getConstantByNameOrNull(llvm::StringRef name) const {
   auto it = nodeValueByName_.find(name);
   if (it == nodeValueByName_.end()) {
     return nullptr;
   }
-  auto *res = llvm::dyn_cast<Constant>(it->second.getNode());
+  NodeValue NV = it->second;
+  auto *res = llvm::dyn_cast<Constant>(NV.getNode());
+  // If name is specified to run in FP16 precision and output type
+  // is FP32 add ConvertTo FP16 node to the original nodeValue and return the
+  // updated nodeValue.
+  // For Constant inputs, convertTofp16 nodes cannot be added
+  // because it won't be Constant anymore and cannot access the tensor
+  // payload. So for Constant inputs change tensor datatype and tensor
+  // payload to fp16.
+  if (res != nullptr && isOpRequestedInFP16Precision(name) &&
+      NV.getElementType() == ElemKind::FloatTy) {
+    auto inTy = mod_.uniqueType(ElemKind::Float16Ty, NV.dims());
+    Tensor &payLoadT = res->getPayloadMutable();
+    payLoadT.convertToType(ElemKind::Float16Ty);
+    res->setType(Storage::OutputIdx, inTy);
+    res->setPayloadType(inTy);
+  }
   return res ? res : nullptr;
```

Comment:
I see this existed before your PR, but it's pretty dumb, can you update while you're here? 🙂 
```suggestion
  return res;
```

Reply:
Done